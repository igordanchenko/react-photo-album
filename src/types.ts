import * as React from "react";

export type LayoutType = "columns" | "rows" | "masonry";

export type PhotoAlbumProps<T extends Photo = Photo> = {
    /** An array of photos to display in the photo album. */
    photos: Array<T>;
    /** Photo album layout type. */
    layout: LayoutType;
    /** A number of columns in the `columns` or `masonry` layout. */
    columns?: ResponsiveParameter;
    /** Spacing between images. */
    spacing?: ResponsiveParameter;
    /** Padding around each image in the photo album. */
    padding?: ResponsiveParameter;
    /** Target row height in the 'rows' layout. */
    targetRowHeight?: ResponsiveParameter;
    /** Additional row constraints in the `rows` layout. */
    rowConstraints?: ResponsiveParameter<RowConstraints>;
    /** Photo album container width at various viewport sizes. */
    sizes?: ResponsiveSizes;
    /** Photo click callback function. */
    onClick?: ClickHandler<T>;
    /** Responsive breakpoints. */
    breakpoints?: number[];
    /** Default container width in SSR. */
    defaultContainerWidth?: number;
    /** Additional HTML attributes to be passed to the rendered elements. */
    componentsProps?: ComponentsPropsParameter;
    /** Custom photo rendering function. */
    renderPhoto?: RenderPhoto<T>;
    /** Custom container rendering function. */
    renderContainer?: RenderContainer;
    /** Custom row container rendering function. */
    renderRowContainer?: RenderRowContainer<T>;
    /** Custom column container rendering function. */
    renderColumnContainer?: RenderColumnContainer<T>;
};

export interface Image {
    /** Image source. */
    src: string;
    /** Image width in pixels. */
    width: number;
    /** Image height in pixels. */
    height: number;
}

export interface Photo extends Image {
    /** Optional `key` attribute. */
    key?: string;
    /** Optional image `alt` attribute. */
    alt?: string;
    /** Optional image `title` attribute. */
    title?: string;
    // TODO v3: remove
    /** @deprecated - use `srcSet` instead */
    images?: Image[];
    /** Optional array of alternative images to be included in the `srcset` attribute. */
    srcSet?: Image[];
}

export type RenderPhotoProps<T extends Photo = Photo> = {
    /** photo object */
    photo: T;
    /** computed photo layout */
    layout: PhotoLayout;
    /** photo album layout options */
    layoutOptions: LayoutOptions<T>;
    /** pre-populated 'img' element attributes */
    imageProps: NonOptional<ImageElementAttributes, "src" | "alt" | "style">;
    /** A callback to render the default photo implementation. If `wrapped` is `true`, the image is styled with `width`
     * and `height` set to 100%. Use this option when rendering image wrapper styled with wrapperStyle. */
    renderDefaultPhoto: RenderFunction<{ wrapped?: boolean } | void>;
    /** CSS styles to properly size image wrapper (i.e. <div> wrapper) */
    wrapperStyle: React.CSSProperties;
};

export type RenderPhoto<T extends Photo = Photo> = RenderFunction<RenderPhotoProps<T>>;

export type ClickHandlerProps<T extends Photo = Photo> = {
    event: React.MouseEvent;
    photo: T;
    index: number;
};

export type ClickHandler<T extends Photo = Photo> = (props: ClickHandlerProps<T>) => void;

export type ResponsiveParameterProvider<T = number> = (containerWidth: number) => T;

export type ResponsiveParameter<T = number> = T | ResponsiveParameterProvider<T>;

export type ResponsiveSizes = {
    /** default size e.g. 100vw or calc(100vw - 200px) */
    size: string;
    /** array of sizes at various breakpoint */
    sizes?: {
        /** viewport size media query e.g. (max-width: 600px)  */
        viewport: string;
        /** photo album width at given viewport size e.g. calc(100vw - 50px) */
        size: string;
    }[];
};

export type PhotoLayout = {
    /** rendered photo width */
    width: number;
    /** rendered photo height */
    height: number;
    /** photo index in the original `photos` array */
    index: number;
    /** photo index in a given row/column */
    photoIndex: number;
    /** number of photos in a given row/column */
    photosCount: number;
};

export type GenericLayoutOptions<T extends Photo = Photo> = {
    /** layout spacing (gaps between photos) */
    spacing: number;
    /** padding around each photo */
    padding: number;
    /** current photo album container width */
    containerWidth: number;
    /** photo click handler */
    onClick?: ClickHandler<T>;
    /** photo album size at various viewport sizes */
    sizes?: ResponsiveSizes;
};

export type RowsLayoutOptions<T extends Photo = Photo> = GenericLayoutOptions<T> & {
    /** layout type */
    layout: Extract<LayoutType, "rows">;
    /** target row height in 'rows' layout */
    targetRowHeight: number;
    /** Additional row constraints */
    rowConstraints?: RowConstraints;
};

export type ColumnsLayoutOptions<T extends Photo = Photo> = GenericLayoutOptions<T> & {
    /** layout type */
    layout: Extract<LayoutType, "columns" | "masonry">;
    /** number of columns in 'columns' or 'masonry' layout */
    columns: number;
};

export type LayoutOptions<T extends Photo = Photo> = ColumnsLayoutOptions<T> | RowsLayoutOptions<T>;

export type RowConstraints = {
    /** minimum number of photos per row in 'rows' layout */
    minPhotos?: number;
    /** maximum number of photos per row in 'rows' layout */
    maxPhotos?: number;
};

export type ComponentsProps = {
    /** Additional HTML attributes to be passed to the outer container `div` element */
    containerProps?: DivElementAttributes;
    /** Additional HTML attributes to be passed to the row container `div` element */
    rowContainerProps?: DivElementAttributes;
    /** Additional HTML attributes to be passed to the column container `div` element */
    columnContainerProps?: DivElementAttributes;
    /** Additional HTML attributes to be passed to the photo `img` element */
    imageProps?: ImageElementAttributes;
};

export type ComponentsPropsParameter = ComponentsProps | ((containerWidth?: number) => ComponentsProps);

export type RenderContainerProps = React.PropsWithChildren<{
    /** layout type */
    layout: LayoutType;
    /** pre-populated default container attributes */
    containerProps: DivElementAttributes;
    /** container ref callback */
    containerRef: React.RefCallback<HTMLDivElement>;
}>;

export type RenderContainer = RenderFunction<RenderContainerProps>;

export type RenderRowContainerProps<T extends Photo = Photo> = React.PropsWithChildren<{
    /** layout options */
    layoutOptions: RowsLayoutOptions<T>;
    /** row number */
    rowIndex: number;
    /** total number of rows */
    rowsCount: number;
    /** pre-populated default row container attributes */
    rowContainerProps: DivElementAttributes;
}>;

export type RenderRowContainer<T extends Photo = Photo> = RenderFunction<RenderRowContainerProps<T>>;

export type RenderColumnContainerProps<T extends Photo = Photo> = React.PropsWithChildren<{
    layoutOptions: ColumnsLayoutOptions<T>;
    /** column number */
    columnIndex: number;
    /** total number of columns */
    columnsCount: number;
    /** sum of spacings and paddings in each column */
    columnsGaps?: number[];
    /** width adjustment ratios of each column */
    columnsRatios?: number[];
    /** pre-populated default column container attributes */
    columnContainerProps: DivElementAttributes;
}>;

export type RenderColumnContainer<T extends Photo = Photo> = RenderFunction<RenderColumnContainerProps<T>>;

export type RenderFunction<T = void> = (props: T) => React.ReactNode;

export type DivElementAttributes = React.HTMLAttributes<HTMLDivElement>;

export type ImageElementAttributes = React.ImgHTMLAttributes<HTMLImageElement>;

export type Optional<T, K extends keyof T> = Partial<Pick<T, K>> & Omit<T, K>;

export type NonOptional<T, K extends keyof T> = Required<Pick<T, K>> & Omit<T, K>;
