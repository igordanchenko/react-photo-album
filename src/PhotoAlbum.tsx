import * as React from "react";

import Layout from "./Layout";
import RowsLayout from "./components/layouts/RowsLayout";
import ColumnsLayout from "./components/layouts/ColumnsLayout";
import MasonryLayout from "./components/layouts/MasonryLayout";
import ContainerRenderer from "./components/renderers/ContainerRenderer";
import resolveResponsiveParameter from "./utils/responsive";
import useLayoutEffect from "./utils/layoutEffect";
import { ColumnsLayoutOptions, Photo, PhotoAlbumProps, RowsLayoutOptions } from "./types";

const resolveLayoutOptions = <T extends Photo>({
    layout,
    onClick,
    viewportWidth,
    containerWidth,
    targetRowHeight,
    columns,
    spacing,
    padding,
}: Omit<PhotoAlbumProps<T>, "photos"> & {
    viewportWidth?: number;
    containerWidth: number;
}): RowsLayoutOptions | ColumnsLayoutOptions => ({
    layout,
    onClick,
    viewportWidth,
    containerWidth,
    columns: resolveResponsiveParameter(columns, containerWidth, [6, 5, 4, 3, 2, 1]),
    spacing: resolveResponsiveParameter(spacing, containerWidth, [20, 16, 12, 8, 4, 2]),
    padding: resolveResponsiveParameter(padding, containerWidth, [0, 0, 0, 0, 0, 0]),
    targetRowHeight: resolveResponsiveParameter(targetRowHeight, containerWidth, [300, 250, 200, 150, 100, 80]),
});

const PhotoAlbum = <T extends Photo>(props: PhotoAlbumProps<T>): JSX.Element => {
    const {
        photos,
        layout,
        columns,
        spacing,
        padding,
        onClick,
        targetRowHeight,
        defaultContainerWidth = 800,
        renderPhoto,
        renderContainer,
        renderRowContainer,
        renderColumnContainer,
        resizeObserverProvider,
        instrumentation,
    } = props;

    const [viewportWidth, setViewportWidth] = React.useState<number>();
    const [containerWidth, setContainerWidth] = React.useState<number>();

    const observerRef = React.useRef<ResizeObserver | null>(null);
    const containerRef = React.useRef<HTMLDivElement | null>(null);

    const updateDimensions = React.useCallback(() => {
        if (typeof window !== "undefined") {
            setViewportWidth(window.innerWidth);
        }
        if (containerRef.current) {
            setContainerWidth(containerRef.current.clientWidth);
        }
    }, []);

    const setContainerRef = React.useCallback((node) => {
        if (observerRef.current && containerRef.current) {
            observerRef.current.unobserve(containerRef.current);
        }

        containerRef.current = node;

        if (node && observerRef.current) {
            observerRef.current.observe(node);
        }
    }, []);

    useLayoutEffect(() => {
        updateDimensions();

        const observerCallback = () => updateDimensions();
        if (typeof ResizeObserver !== "undefined") {
            observerRef.current = new ResizeObserver(observerCallback);
        } else if (resizeObserverProvider) {
            observerRef.current = resizeObserverProvider(observerCallback);
        }

        if (observerRef.current && containerRef.current) {
            observerRef.current.observe(containerRef.current);
        }

        return () => {
            if (observerRef.current) {
                observerRef.current.disconnect();
                observerRef.current = null;
            }
        };
    }, [updateDimensions, resizeObserverProvider]);

    const layoutOptions = resolveLayoutOptions({
        containerWidth: containerWidth || defaultContainerWidth,
        viewportWidth,
        layout,
        onClick,
        spacing,
        padding,
        columns,
        targetRowHeight,
    });

    const commonLayoutProps = { photos, renderPhoto, instrumentation };

    return (
        <ContainerRenderer ref={setContainerRef} layoutOptions={layoutOptions} renderContainer={renderContainer}>
            {layout === Layout.Rows ? (
                <RowsLayout<T>
                    layoutOptions={layoutOptions as RowsLayoutOptions}
                    renderRowContainer={renderRowContainer}
                    {...commonLayoutProps}
                />
            ) : layout === Layout.Columns ? (
                <ColumnsLayout
                    layoutOptions={layoutOptions as ColumnsLayoutOptions}
                    renderColumnContainer={renderColumnContainer}
                    {...commonLayoutProps}
                />
            ) : (
                <MasonryLayout
                    layoutOptions={layoutOptions as ColumnsLayoutOptions}
                    renderColumnContainer={renderColumnContainer}
                    {...commonLayoutProps}
                />
            )}
        </ContainerRenderer>
    );
};

export default PhotoAlbum;
