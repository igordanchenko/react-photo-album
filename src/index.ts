export * from "./types";

export { default } from "./client/aggregate";
export { default as RowsPhotoAlbum } from "./client/rows";
export { default as ColumnsPhotoAlbum } from "./client/columns";
export { default as MasonryPhotoAlbum } from "./client/masonry";

// experimental exports (no semver coverage)
export { default as UnstableStaticPhotoAlbum } from "./static";
export type { StaticPhotoAlbumProps as UnstableStaticPhotoAlbumProps } from "./static";

export { default as unstable_computeRowsLayout } from "./layouts/rows";
export { default as unstable_computeColumnsLayout } from "./layouts/columns";
export { default as unstable_computeMasonryLayout } from "./layouts/masonry";
