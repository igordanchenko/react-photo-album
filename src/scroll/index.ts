export type { InfiniteScrollProps as UnstableInfiniteScrollProps } from "./InfiniteScroll";
export { default as UnstableInfiniteScroll } from "./InfiniteScroll";
