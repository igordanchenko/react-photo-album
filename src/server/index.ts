export type { ServerPhotoAlbumProps as UnstableServerPhotoAlbumProps } from "./ServerPhotoAlbum";
export { default as UnstableServerPhotoAlbum } from "./ServerPhotoAlbum";
