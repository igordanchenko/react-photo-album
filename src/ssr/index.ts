export type { SSRProps as UnstableSSRProps } from "./SSR";
export { default as UnstableSSR } from "./SSR";
