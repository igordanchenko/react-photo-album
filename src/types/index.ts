import * as React from "react";

export type LayoutType = "columns" | "rows" | "masonry";

export type ClickHandler<T extends Photo = Photo> = (event: React.MouseEvent, photo: T, index: number) => void;

export type ResponsiveParameterProvider<T = number> = (containerWidth: number) => T;

export type ResponsiveParameter<T = number> = T | ResponsiveParameterProvider<T>;

export type ResponsiveSizes = {
    /** default size e.g. 100vw or calc(100vw - 200px) */
    size: string;
    /** array of sizes at various breakpoint */
    sizes?: {
        /** viewport size media query e.g. (max-width: 600px)  */
        viewport: string;
        /** photo album width at given viewport size e.g. calc(100vw - 50px) */
        size: string;
    }[];
};

export interface Image {
    /** Image source. */
    src: string;
    /** Image width in pixels. */
    width: number;
    /** Image height in pixels. */
    height: number;
}

export interface Photo extends Image {
    /** Optional `key` attribute. */
    key?: string;
    /** Optional image `alt` attribute. */
    alt?: string;
    /** Optional image `title` attribute. */
    title?: string;
    /** Optional array of alternative images to be included in the `srcset` attribute. */
    images?: Image[];
}

export type PhotoLayout = {
    /** photo width */
    width: number;
    /** photo height */
    height: number;
    /** photo index in the original `photos` array */
    index: number;
    /** photo index in a given row/column */
    photoIndex: number;
    /** number of photos in a given row/column */
    photosCount: number;
};

export type PhotoProps<T extends Photo = Photo> = {
    /** photo object */
    photo: T;
    /** computed photo layout */
    layout: PhotoLayout;
    /** photo album layout options */
    layoutOptions: LayoutOptions<T>;
    /** pre-populated 'img' element attributes */
    imageProps: React.ImgHTMLAttributes<HTMLImageElement> &
        Required<Pick<React.ImgHTMLAttributes<HTMLImageElement>, "src" | "alt" | "style">>;
    /** A callback to render the default photo implementation. If `wrapped` is `true`, the image will be styled
     * with `width` and `height` set to 100%. Use this option when you already sized image wrapper with image
     * dimensions (see `wrapperStyle`) */
    renderDefaultPhoto: (options?: { wrapped?: boolean }) => React.ReactNode;
    /** CSS styles to properly size image wrapper (i.e. <div> wrapper) */
    wrapperStyle: React.CSSProperties;
};

export type PhotoAlbumProps<T extends Photo = Photo> = {
    /** An array of photos to display in the photo album. */
    photos: Array<T>;
    /** Photo album layout type. */
    layout: LayoutType;
    /** A number of columns in the `columns` or `masonry` layout. */
    columns?: ResponsiveParameter;
    /** Spacing between images. */
    spacing?: ResponsiveParameter;
    /** Padding around each image in the photo album. */
    padding?: ResponsiveParameter;
    /** Target row height in the 'rows' layout. */
    targetRowHeight?: ResponsiveParameter;
    /** Additional row constraints */
    rowConstraints?: ResponsiveParameter<RowConstraints>;
    /** Additional columnn constraints */
    columnConstraints?: ResponsiveParameter<ColumnConstraints>;
    /** Photo album width at various viewport sizes. */
    sizes?: ResponsiveSizes;
    /** Photo click callback function. */
    onClick?: ClickHandler<T>;
    /** Responsive breakpoints */
    breakpoints?: number[];
    /** Default container width to be used in the server-side render. */
    defaultContainerWidth?: number;
    /** Additional HTML attributes to be passed to the rendered elements */
    componentsProps?: ComponentsPropsParameter;
    /** Custom photo rendering function. */
    renderPhoto?: RenderPhoto<T>;
    /** Custom container rendering function. */
    renderContainer?: RenderContainer<T>;
    /** Custom row container rendering function. */
    renderRowContainer?: RenderRowContainer<T>;
    /** Custom column container rendering function. */
    renderColumnContainer?: RenderColumnContainer<T>;
    /** Internal instrumentation - use on your own risk. */
    instrumentation?: Instrumentation;
};

export type RenderPhoto<T extends Photo = Photo> = (props: PhotoProps<T>) => React.ReactNode;

export type GenericLayoutOptions<T extends Photo = Photo> = {
    /** layout spacing (gaps between photos) */
    spacing: number;
    /** padding around each photo */
    padding: number;
    /** current photo album container width */
    containerWidth: number;
    /** current viewport width */
    viewportWidth?: number;
    /** photo click handler */
    onClick?: ClickHandler<T>;
    /** photo album size at various viewport sizes */
    sizes?: ResponsiveSizes;
};

export type RowsLayoutOptions<T extends Photo = Photo> = GenericLayoutOptions<T> & {
    /** layout type */
    layout: Extract<LayoutType, "rows">;
    /** target row height in 'rows' layout */
    targetRowHeight: number;
    /** Additional row constraints */
    rowConstraints?: RowConstraints;
};

export type ColumnsLayoutOptions<T extends Photo = Photo> = GenericLayoutOptions<T> & {
    /** layout type */
    layout: Extract<LayoutType, "columns" | "masonry">;
    /** number of columns in 'columns' or 'masonry' layout */
    columns: number;
    /** Additional column constraints */
    columnConstraints?: ColumnConstraints;
};

export type LayoutOptions<T extends Photo = Photo> = ColumnsLayoutOptions<T> | RowsLayoutOptions<T>;

export type ComponentsProps = {
    /** Additional HTML attributes to be passed to the outer container `div` element */
    containerProps?: React.HTMLAttributes<HTMLDivElement>;
    /** Additional HTML attributes to be passed to the row container `div` element */
    rowContainerProps?: React.HTMLAttributes<HTMLDivElement>;
    /** Additional HTML attributes to be passed to the column container `div` element */
    columnContainerProps?: React.HTMLAttributes<HTMLDivElement>;
    /** Additional HTML attributes to be passed to the photo `img` element */
    imageProps?: React.HTMLAttributes<HTMLImageElement>;
};

export type ComponentsPropsParameter = ComponentsProps | ((containerWidth: number) => ComponentsProps);

export type ContainerProps<T extends Photo = Photo> = {
    /** layout options */
    layoutOptions: LayoutOptions<T>;
    /** pre-populated default container attributes */
    containerProps: React.HTMLAttributes<HTMLDivElement>;
};

export type RenderContainerProps<T extends Photo = Photo> = React.PropsWithChildren<ContainerProps<T>> & {
    containerRef?: React.RefCallback<HTMLDivElement>;
};

/** ForwardRefExoticComponent (forwardRef) variant is deprecated and will be removed in the next major release */
export type RenderContainer<T extends Photo = Photo> = (props: RenderContainerProps<T>) => React.ReactNode;

export type RowContainerProps<T extends Photo = Photo> = {
    /** layout options */
    layoutOptions: RowsLayoutOptions<T>;
    /** row number */
    rowIndex: number;
    /** total number of rows */
    rowsCount: number;
    /** pre-populated default row container attributes */
    rowContainerProps: React.HTMLAttributes<HTMLDivElement>;
};

export type RenderRowContainer<T extends Photo = Photo> = (
    props: React.PropsWithChildren<RowContainerProps<T>>
) => React.ReactNode;

export type ColumnContainerProps<T extends Photo = Photo> = {
    layoutOptions: ColumnsLayoutOptions<T>;
    /** column number */
    columnIndex: number;
    /** total number of columns */
    columnsCount: number;
    /** sum of spacings and paddings in each column */
    columnsGaps?: number[];
    /** width adjustment ratios of each column */
    columnsRatios?: number[];
    /** pre-populated default column container attributes */
    columnContainerProps: React.HTMLAttributes<HTMLDivElement>;
};

export type RenderColumnContainer<T extends Photo = Photo> = (
    props: React.PropsWithChildren<ColumnContainerProps<T>>
) => React.ReactNode;

export type RowConstraints = {
    /** minimum number of photos per row in 'rows' layout */
    minPhotos?: number;
    /** maximum number of photos per row in 'rows' layout */
    maxPhotos?: number;
};

export type ColumnConstraints = {
    /** minimum number of columns in 'masonry' or 'columns' layout when there isn't enough photos to fill all the columns */
    minColumns?: number;
};

/** internal instrumentation for research and performance testing purposes, subject to change without notice */
export type Instrumentation = {
    fullGraphSearch?: boolean;
    onStartLayoutComputation?: () => void;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    onFinishLayoutComputation?: (...params: any) => void;
};
