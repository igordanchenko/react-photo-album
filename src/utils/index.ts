export { default as clsx } from "./clsx";
export * from "./css";
export { default as ratio } from "./ratio";
export * from "./responsive";
export { default as round } from "./round";
export { default as srcSetAndSizes } from "./sizes";
