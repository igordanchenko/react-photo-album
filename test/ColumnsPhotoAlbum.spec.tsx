import { ColumnsPhotoAlbum } from "../src";
import { render, renderAndMatchSnapshot } from "./test-utils";
import photos from "./photos";

describe("ColumnsPhotoAlbum", () => {
  it("renders without crashing", () => {
    renderAndMatchSnapshot(<ColumnsPhotoAlbum photos={[]} />);
    renderAndMatchSnapshot(<ColumnsPhotoAlbum photos={photos} />);
  });

  it("supports default responsive columns prop", () => {
    const { getTracks } = render(<ColumnsPhotoAlbum photos={photos} />);

    window.resizeTo(1400, 600);
    expect(getTracks().length).toBe(5);

    window.resizeTo(800, 600);
    expect(getTracks().length).toBe(4);

    window.resizeTo(500, 600);
    expect(getTracks().length).toBe(3);

    window.resizeTo(200, 600);
    expect(getTracks().length).toBe(2);
  });

  it("fills first N available columns", () => {
    const { getTracks } = render(<ColumnsPhotoAlbum photos={photos.slice(0, 2)} columns={5} />);
    expect(getTracks().length).toBe(5);
  });

  it("handles impossible layout", () => {
    window.resizeTo(400, 600);
    const { getTracks } = render(<ColumnsPhotoAlbum photos={photos} spacing={20} padding={50} columns={5} />);
    expect(getTracks().length).toBe(3);

    window.resizeTo(80, 600);
    expect(getTracks().length).toBe(0);
  });
});
