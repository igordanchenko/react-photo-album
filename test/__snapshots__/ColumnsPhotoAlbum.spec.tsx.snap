// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ColumnsPhotoAlbum > renders without crashing 1`] = `
<DocumentFragment>
  <div
    aria-label="Photo album"
    class="react-photo-album react-photo-album--columns"
    role="group"
    style="--react-photo-album--spacing: 15; --react-photo-album--padding: 0; --react-photo-album--container-width: 1024; --react-photo-album--columns: 4; --react-photo-album--total-ratio: 4;"
  >
    <div
      class="react-photo-album--track"
      style="--react-photo-album--track-size: 0; --react-photo-album--adjusted-gaps: 0; --react-photo-album--column-ratio: 1;"
    />
    <div
      class="react-photo-album--track"
      style="--react-photo-album--track-size: 0; --react-photo-album--adjusted-gaps: 0; --react-photo-album--column-ratio: 1;"
    />
    <div
      class="react-photo-album--track"
      style="--react-photo-album--track-size: 0; --react-photo-album--adjusted-gaps: 0; --react-photo-album--column-ratio: 1;"
    />
    <div
      class="react-photo-album--track"
      style="--react-photo-album--track-size: 0; --react-photo-album--adjusted-gaps: 0; --react-photo-album--column-ratio: 1;"
    />
  </div>
</DocumentFragment>
`;

exports[`ColumnsPhotoAlbum > renders without crashing 2`] = `
<DocumentFragment>
  <div
    aria-label="Photo album"
    class="react-photo-album react-photo-album--columns"
    role="group"
    style="--react-photo-album--spacing: 15; --react-photo-album--padding: 0; --react-photo-album--container-width: 1024; --react-photo-album--columns: 4; --react-photo-album--total-ratio: 0.81179;"
  >
    <div
      class="react-photo-album--track"
      style="--react-photo-album--track-size: 5; --react-photo-album--adjusted-gaps: -3.22114; --react-photo-album--column-ratio: 0.19704;"
    >
      <div
        class="react-photo-album--photo"
        style="--react-photo-album--photo-width: 238.414; --react-photo-album--photo-height: 176.603;"
      >
        <img
          class="react-photo-album--image"
          decoding="async"
          loading="lazy"
          sizes="24vw"
          src="image01.jpg"
        />
      </div>
      <div
        class="react-photo-album--photo"
        style="--react-photo-album--photo-width: 238.414; --react-photo-album--photo-height: 357.62;"
      >
        <img
          class="react-photo-album--image"
          decoding="async"
          loading="lazy"
          sizes="24vw"
          src="image02.jpg"
        />
      </div>
      <div
        class="react-photo-album--photo"
        style="--react-photo-album--photo-width: 238.414; --react-photo-album--photo-height: 158.942;"
      >
        <img
          class="react-photo-album--image"
          decoding="async"
          loading="lazy"
          sizes="24vw"
          src="image03.jpg"
        />
      </div>
      <div
        class="react-photo-album--photo"
        style="--react-photo-album--photo-width: 238.414; --react-photo-album--photo-height: 159.163;"
      >
        <img
          class="react-photo-album--image"
          decoding="async"
          loading="lazy"
          sizes="24vw"
          src="image04.jpg"
        />
      </div>
      <div
        class="react-photo-album--photo"
        style="--react-photo-album--photo-width: 238.414; --react-photo-album--photo-height: 357.62;"
      >
        <img
          class="react-photo-album--image"
          decoding="async"
          loading="lazy"
          sizes="24vw"
          src="image05.jpg"
        />
      </div>
    </div>
    <div
      class="react-photo-album--track"
      style="--react-photo-album--track-size: 6; --react-photo-album--adjusted-gaps: 8.95567; --react-photo-album--column-ratio: 0.2205;"
    >
      <div
        class="react-photo-album--photo"
        style="--react-photo-album--photo-width: 263.484; --react-photo-album--photo-height: 148.088;"
      >
        <img
          class="react-photo-album--image"
          decoding="async"
          loading="lazy"
          sizes="26vw"
          src="image06.jpg"
        />
      </div>
      <div
        class="react-photo-album--photo"
        style="--react-photo-album--photo-width: 263.484; --react-photo-album--photo-height: 148.332;"
      >
        <img
          class="react-photo-album--image"
          decoding="async"
          loading="lazy"
          sizes="26vw"
          src="image07.jpg"
        />
      </div>
      <div
        class="react-photo-album--photo"
        style="--react-photo-album--photo-width: 263.484; --react-photo-album--photo-height: 175.656;"
      >
        <img
          class="react-photo-album--image"
          decoding="async"
          loading="lazy"
          sizes="26vw"
          src="image08.jpg"
        />
      </div>
      <div
        class="react-photo-album--photo"
        style="--react-photo-album--photo-width: 263.484; --react-photo-album--photo-height: 377.905;"
      >
        <img
          class="react-photo-album--image"
          decoding="async"
          loading="lazy"
          sizes="26vw"
          src="image09.jpg"
        />
      </div>
      <div
        class="react-photo-album--photo"
        style="--react-photo-album--photo-width: 263.484; --react-photo-album--photo-height: 175.656;"
      >
        <img
          class="react-photo-album--image"
          decoding="async"
          loading="lazy"
          sizes="26vw"
          src="image10.jpg"
        />
      </div>
      <div
        class="react-photo-album--photo"
        style="--react-photo-album--photo-width: 263.484; --react-photo-album--photo-height: 169.313;"
      >
        <img
          class="react-photo-album--image"
          decoding="async"
          loading="lazy"
          sizes="26vw"
          src="image11.jpg"
        />
      </div>
    </div>
    <div
      class="react-photo-album--track"
      style="--react-photo-album--track-size: 4; --react-photo-album--adjusted-gaps: -15.39795; --react-photo-album--column-ratio: 0.2;"
    >
      <div
        class="react-photo-album--photo"
        style="--react-photo-album--photo-width: 244.99; --react-photo-album--photo-height: 367.485;"
      >
        <img
          class="react-photo-album--image"
          decoding="async"
          loading="lazy"
          sizes="24vw"
          src="image12.jpg"
        />
      </div>
      <div
        class="react-photo-album--photo"
        style="--react-photo-album--photo-width: 244.99; --react-photo-album--photo-height: 163.327;"
      >
        <img
          class="react-photo-album--image"
          decoding="async"
          loading="lazy"
          sizes="24vw"
          src="image13.jpg"
        />
      </div>
      <div
        class="react-photo-album--photo"
        style="--react-photo-album--photo-width: 244.99; --react-photo-album--photo-height: 326.653;"
      >
        <img
          class="react-photo-album--image"
          decoding="async"
          loading="lazy"
          sizes="24vw"
          src="image14.jpg"
        />
      </div>
      <div
        class="react-photo-album--photo"
        style="--react-photo-album--photo-width: 244.99; --react-photo-album--photo-height: 367.485;"
      >
        <img
          class="react-photo-album--image"
          decoding="async"
          loading="lazy"
          sizes="24vw"
          src="image15.jpg"
        />
      </div>
    </div>
    <div
      class="react-photo-album--track"
      style="--react-photo-album--track-size: 6; --react-photo-album--adjusted-gaps: 8.95567; --react-photo-album--column-ratio: 0.19424;"
    >
      <div
        class="react-photo-album--photo"
        style="--react-photo-album--photo-width: 232.112; --react-photo-album--photo-height: 174.084;"
      >
        <img
          class="react-photo-album--image"
          decoding="async"
          loading="lazy"
          sizes="23vw"
          src="image16.jpg"
        />
      </div>
      <div
        class="react-photo-album--photo"
        style="--react-photo-album--photo-width: 232.112; --react-photo-album--photo-height: 348.169;"
      >
        <img
          class="react-photo-album--image"
          decoding="async"
          loading="lazy"
          sizes="23vw"
          src="image17.jpg"
        />
      </div>
      <div
        class="react-photo-album--photo"
        style="--react-photo-album--photo-width: 232.112; --react-photo-album--photo-height: 34.387;"
      >
        <img
          class="react-photo-album--image"
          decoding="async"
          loading="lazy"
          sizes="23vw"
          src="image18.jpg"
        />
      </div>
      <div
        class="react-photo-album--photo"
        style="--react-photo-album--photo-width: 232.112; --react-photo-album--photo-height: 174.084;"
      >
        <img
          class="react-photo-album--image"
          decoding="async"
          loading="lazy"
          sizes="23vw"
          src="image19.jpg"
        />
      </div>
      <div
        class="react-photo-album--photo"
        style="--react-photo-album--photo-width: 232.112; --react-photo-album--photo-height: 154.742;"
      >
        <img
          class="react-photo-album--image"
          decoding="async"
          loading="lazy"
          sizes="23vw"
          src="image20.jpg"
        />
      </div>
      <div
        class="react-photo-album--photo"
        style="--react-photo-album--photo-width: 232.112; --react-photo-album--photo-height: 309.483;"
      >
        <img
          class="react-photo-album--image"
          decoding="async"
          loading="lazy"
          sizes="23vw"
          src="image21.jpg"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
